### PR TITLE
perf: Optimize files, folder and extension lookups in Context

### LINKED ISSUE #852: Replace the use of Vector<String> with HashSet<String> for Context dir_files
## Feature Request

As suggested by @davidkna in #788, we should probably be using a `HashSet` to store the files, folders, and extensions in the current directory within `Context` for constant lookup times, instead of linear lookup times, increasing as we search for more files.

#### Is your feature request related to a problem? Please describe.

Lookup times are increasing as more files, folders, and extensions are looked for by various modules.

#### Describe the solution you'd like

Replace the use of `Vector<String>` with a `HashSet<String>` for `Context.dir_files`

### PR DESCRIPTION
**Pull request by samwho**
#### Description

Changes the `Vec<PathBuf>` containing all files and folders in context.rs to be a new struct called `DirContents` that optimises for lookup time on various relevant entities, e.g. files, folders and file names.

#### Motivation and Context

Closes #852.

#### Types of changes 
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
 
#### How Has This Been Tested?
- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
- [ ] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

**Comment by samwho** [comment]
Will work on getting all of the tests to pass over the next week, time allowing. I thought I had run all of the tests 🤔

**Comment by samwho** [comment]
I _think_ this is ready now. The last MacOS test seems to be taking a very long time, but the rest passed.

**Comment by chipbuster** [comment]
I'll want to run more benchmarks on various systems in various configurations, but the single data point I have so far makes this look amazing.

On my home PC (quad-core 2013 CPU, SATA SSD), this change drops the render time for a 100k-file directory (with no special files) from 185ms to 34.5ms.

**Comment by samwho** [comment]
I was going to ask if we have some accepted set of benchmark we use for changes like this. Intuitively it makes sense for this to have a positive effect on large projects, but you're absolutely right to get some numbers. 😄

**Comment by chipbuster** [comment]
We've been trying to get a standard set of benchmarks written up and included with `cargo bench`, but as of right now that's more of a "we should do that someday" item than a reality.

File: src/context.rs [hunk 1]
```diff
@@ -139,26 +138,101 @@ impl<'a> Context<'a> {
...
+    // HashSet of all file names, e.g. the last section without any folders, as strings.
+    file_names: HashSet<String>,
+    // HashSet of all folders, relative to the base directory given at construction.
+    folders: HashSet<PathBuf>,
+    // HashSet of all extensions found, without dots, e.g. "js" instead of ".js".
+    extensions: HashSet<String>,
+}
+
+impl DirContents {
+    fn from_path(base: &PathBuf) -> Result<Self, std::io::Error> {
```
Comment (matchai) [inline src/context.rs]: It looks like this function is unused. Perhaps we should remove it?

Comment (samwho) [reply src/context.rs]: It's used in the tests to avoid having to pass in a bogus Duration. Is that okay?

**Comment by Akoneto9** [comment]
Check mind too

**Pull request merged**